src/js/store/middleware/feature/form.js

```diff
@@ -226,6 +265,46 @@ const handleClear = (store, next) => {
   ])
 }
 
+const handlePersistInLocalStorage = (store) => {
+  const state = store.getState()
+  const key = localStorageKey(state)
+  const storageState = {
+    schema: state.form.schema,
+    fields: state.form.fields,
+  }
+  window.localStorage.setItem(key, JSON.stringify(storageState))
+}
+
+const handleClearLocalStorage = (store) => {
+  const state = store.getState()
+  const key = localStorageKey(state)
+  window.localStorage.removeItem(key)
+}
+
+const handleAcceptLocalStorage = (store, next) => {
+  const state = store.getState()
+  const key = localStorageKey(state)
+  let storageState = { fields: {} }
+  try {
+    storageState = JSON.parse(window.localStorage.getItem(key))
+  } catch(error) {
+    console.error(error)
+  }
+  Object.entries(storageState.fields).forEach(([fieldId, value]) => {
+    store.dispatch(inputField({ fieldId, value }))
+  })
+  next(setAskingToRestore({ state: false, feature: FORM }))
```

Comment:
When a user accepts the data found in local storage (only an option when some data is found in the first place _and_ the locally stored schema matches the newly loaded schema), we parse the field values stored in local storage and dispatch an `INPUT_FIELD` action for each field value. This spoofs actual user input (i.e., it's more high-level than the `SET_FIELD` action that just sets the field state in the store) and triggers form validation, etc. The end result should be that the form ends up exactly as it would if the user had manually re-entered all of the fields through the UI.

Reply:
The other thing we do here is check if the index was loaded at the end of the user's previously interrupted session. If it was, update the UI to reflect that so that we render the complete form with all of the user's work.

Note that in practice this should always be the case because we don't actually treat an indexed form as "dirty" and start persisting a user's work in local storage until _after_ they've loaded an index. In theory that could change in the future and persisting the UI state seems like the cleanest thing at the moment.